### ISSUE TITLE
Server freezes after running for a while

### ISSUE THREAD

**Issue by ghost** [input-leap/input-leap#1248]
> This issue has been migrated from old Barrier Github repository https://github.com/debauchee/barrier/issues/1248
> 
> Issue created on: 2021-08-15 by @roranicus01
> Issue last updated on: 2021-09-09

**Describe the bug**
After gaming on the Windows 10 client for, say, twenty minutes or so, my Debian 10 server will freeze. I lose control of both keyboard and mouse, am unable to switch machine, and the only fix is a hard reset.

**To Reproduce**

It happens whenever I only use the client computer for more than a few minutes. Eventually, the server freezes. 

**Expected behavior**
The ability to use the client computer without the server ever freezing or slowing down.

**Screenshots**

Please do not link to image hosting sites, as these can be ephemeral. Instead, attach them to the issue. 

**Desktop (please complete the following information):**

 - OS: Debian 10
- Barrier version 2.3.3

**Additional context**

I've only been using Barrier for a day, and am new to the program. 

---

> Commented on: 2021-08-17 by @roranicus01

After experimenting some more, I found that the server doesn't actually freeze. Rather, I lose control of my mouse and keyboard completely. Unplugging and replugging them doesn't work. The only way I can fix it is a hard reboot. 

---

> Commented on: 2021-09-09 by @denego

I have a similar issue and know how to reproduce.

**Describe the bug**
The Ubuntu machine running barrier server loses the mouse and keyboard. The machine is keep running, I can ssh into it and when I run `showkey` I can see the keyboard events, but the system just doesn't respond to the events. Restarting display-manager (`systemctl restart display-manager`) fixes everything.

**To Reproduce**
Run server 2.3.3 or 2.3.2 on Ubuntu and a client 2.3.2 on MacOS M1. On Ubuntu copy an image into the clipboard. The funny part is the image must be big enough, like 600x600. Start moving the cursor to the screen edge where the client is. And once the cursor jumps to the client, the server loses the mouse and keyboard, client disconnects.

**Expected behavior**
The server should gracefully handle the error without breaking the system.

**Logs**
```
[2021-09-09T13:48:12] DEBUG2: event: MotionNotify 3,621
[2021-09-09T13:48:12] DEBUG2: event: MotionNotify 0,621
[2021-09-09T13:48:12] DEBUG2: find neighbor on left of "tui"
[2021-09-09T13:48:12] DEBUG2: "Dennis-P.local" is on left of "tui" at 0.575463
[2021-09-09T13:48:12] DEBUG1: try to leave "tui" on left
[2021-09-09T13:48:12] DEBUG2: mapping state: 0
[2021-09-09T13:48:12] INFO: switch from "tui" to "Dennis-P.local" at 1439,517
[2021-09-09T13:48:12] INFO: leaving screen
[2021-09-09T13:48:12] DEBUG2: grabbed keyboard
[2021-09-09T13:48:12] DEBUG1: grabbed pointer and keyboard
[2021-09-09T13:48:12] DEBUG2: warped to 960,540
[2021-09-09T13:48:12] DEBUG2: mapping state: 0
[2021-09-09T13:48:12] DEBUG1: modifiers on update: 0x0000
[2021-09-09T13:48:12] DEBUG: open clipboard 0
[2021-09-09T13:48:12] DEBUG1: locked motif clipboard
[2021-09-09T13:48:12] DEBUG2: can't read property 591 on window 0x00000242
[2021-09-09T13:48:12] DEBUG1: motif does not own clipboard
[2021-09-09T13:48:12] DEBUG1: unlocked motif clipboard
[2021-09-09T13:48:12] DEBUG1: request selection=CLIPBOARD (507), target=TIMESTAMP (519), window=3a00004
[2021-09-09T13:48:12] DEBUG2: read property 539 on window 0x03a00004: bytes=4
[2021-09-09T13:48:12] DEBUG1:   target INTEGER (19)
[2021-09-09T13:48:12] DEBUG1:   got data, 4 bytes
[2021-09-09T13:48:12] DEBUG1: request succeeded after 0.000003s
[2021-09-09T13:48:12] DEBUG1: got ICCCM time 8849420
[2021-09-09T13:48:12] DEBUG: ICCCM fill clipboard 0
[2021-09-09T13:48:12] DEBUG1: request selection=CLIPBOARD (507), target=TARGETS (518), window=3a00004
[2021-09-09T13:48:12] DEBUG2: read property 539 on window 0x03a00004: bytes=40
[2021-09-09T13:48:12] DEBUG1:   target ATOM (4)
[2021-09-09T13:48:12] DEBUG1:   got data, 40 bytes
[2021-09-09T13:48:12] DEBUG1: request succeeded after 0.000006s
[2021-09-09T13:48:12] DEBUG:   available targets: TARGETS (518), MULTIPLE (514), image/tiff (620), image/jpeg (619), image/x-MS-bmp (610)
[2021-09-09T13:48:12] DEBUG1: request selection=CLIPBOARD (507), target=text/html (593), window=3a00004
[2021-09-09T13:48:12] DEBUG2: can't read property 539 on window 0x03a00004
[2021-09-09T13:48:12] DEBUG1: request failed after 0.000007s
[2021-09-09T13:48:12] DEBUG1: can't get data for selection target text/html (593)
[2021-09-09T13:48:12] DEBUG1:   no data for target text/html (593)
[2021-09-09T13:48:12] DEBUG1: request selection=CLIPBOARD (507), target=application/x-moz-nativehtml (594), window=3a00004
[2021-09-09T13:48:12] DEBUG2: can't read property 539 on window 0x03a00004
[2021-09-09T13:48:12] DEBUG1: request failed after 0.000009s
[2021-09-09T13:48:12] DEBUG1: can't get data for selection target application/x-moz-nativehtml (594)
[2021-09-09T13:48:12] DEBUG1:   no data for target application/x-moz-nativehtml (594)
[2021-09-09T13:48:12] DEBUG1: request selection=CLIPBOARD (507), target=image/bmp (595), window=3a00004
[2021-09-09T13:48:12] DEBUG2: read property 539 on window 0x03a00004: bytes=4
[2021-09-09T13:48:21] DEBUG2: reading secure socket
[2021-09-09T13:48:21] DEBUG: ssl connection closed
[2021-09-09T13:48:21] DEBUG1: registered event type stopRetry as 49
[2021-09-09T13:48:21] DEBUG1: registered event type disconnected as 50
```

**Desktop (please complete the following information):**
OS: Ubuntu 18 Mate, i3wm
Barrier: 2.3.3 (Sept 5, 2021)

**Additional context**
To work around this I had to disable clipboard sharing, which is a little inconvinient.

**Comment by vancluever** [comment]
I think I have a full repro of this issue:

* Start server and a client
* *Without touching input on the server*, copy something on the client (ie: select some text, copy something, basically trigger an update).
* Make the very next thing you do be moving focus to the client.

This should trigger the freeze. If you've compiled w/debug symbols, you probably will not make it to this point due to the assertion failure that happens as explained in #1593, you'll dump core before the bug can trigger.

This honestly looks like it's existed for a _long_ time. The code here is pretty old and probably does not come up that much due to the super edge case nature it happens in (and also clearly on some platforms the assertion possibly guarding against it by just causing the server to restart). The main issue is that the clipboard events seem to take some time to come through on some platforms (ie: MacOS), like maybe a quarter to half-second, and as a result the possibility exists that you could mouse over to another screen before it fires. Otherwise the assertion could probably just be first-classed so that there's no possibility of anything ever sending over clipboard events when it's not the active screen/client.

**Comment by shymega** [comment]
Yeah, this looks like a long-standing bug. I can't merge these linked PRs until @p12tic has a look, because I'm not familiar with the codebase in depth, and am more versed with Rust.
